### ISSUE TITLE
Automatically generate ObjectResource case classes from CRDs

### ISSUE THREAD

**Issue by an-tex** [hagay3/skuber#193]
**Is your feature request related to a problem? Please describe.**
You can already create ObjectResource case classes manually [as described here](https://skuber.co/#/?id=custom-resource). These days it's common to have many operators and their CRDs installed in a k8s cluster which makes it quite cumbersome to manually create those often deeply nested structures.

**Describe the solution you'd like**
To me it seems the only easy way out (at least from the users point of view) is to have some kind of code generation in place which translates those YAML based CRDs ([EventBus example](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/base/crds/argoproj.io_eventbus.yaml)) into simple case classes. I quite like the approach [ScalablyTyped](https://github.com/ScalablyTyped/Converter) takes. It's a sbt plugin taking TypeScript and converting it to Scala.JS facades, such that you can directly use JavaScript libraries in a typed way directly from Scala.JS. This all happens at build time so there's no need to commit the generated classes into a VCS. For this to work you'd have to define all the CRD dependency URLs you have in your build file (similar to the NPM dependencies for ScalablyTyped). In theory just linking to (the often existent) `manifests/kustomization.yaml` of a k8s application ([EventBus example](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/cluster-install/kustomization.yaml)) be sufficient.

**Additional context**
I'm currently using kustomize with ArgoCD to achieve GitOPs principles. It works great but with increasing cluster complexity kustomize becomes harder and harder to maintain. Having the CRD generation feature in place it should be possible to describe the whole k8s cluster using Skuber, export it as a big yaml and feed it back into ArgoCD. IaaC using typed scala code would be so awesome :)

**Comment by hagay3** [comment]
Hi @an-tex 
Thank you for opening the issue and showing interest in skuber.

I think you are talking about dynamic client.
This exact client already implemented in other k8s clients( python,java)
This is what are you referring to?

https://caiorcferreira.github.io/post/the-kubernetes-dynamic-client/

Would you like to collaborate and work on this feature together?

**Comment by an-tex** [comment]
Hi @hagay3 ,

thanks for the swift reply! 

The dynamic client seems to be a rather untyped (except for the core k8s types) construct. I'm talking about a typed way to use CRDs defined by external projects like the Argo EventBus CRDs. 

There is a [code generation approach in the kubernetes java client](https://github.com/kubernetes-client/java/blob/master/docs/generate-model-from-third-party-resources.md) though. This goes in the right direction, even though I'm wondering about the need to spawn a kubernetes cluster "just" to apply the CRDs and get the equivalent OpenAPI specs to generate the java classes. It might be the way to go though I'm not enough of an expert to judge here.

I'd be happy to be part of the discussion but I'm afraid it'll be tight to fit in any actual code contribution. The thought about the use case just came up and I thought I'd rather bring it up here before it gets lost again ;)

**Comment by hagay3** [comment]
Hi @an-tex 
Yes that's also an option, generating classes from yamls related to OpenAPI Integration #180.
I think it will be much more complicated to use third-party resources with such integration because it will require you to create those yamls definitions ([example](https://gist.githubusercontent.com/yue9944882/266fee8e95c2f15a93778263633e72ed/raw/be12c13379eeed13d2532cb65da61fffb19ee3e7/crontab-crd.yaml)), generate the case class with "class generator" and use it with skuber.
Also, it will couple dynamic client feature with the development of OpenAPI integration.

Using dynamic client will allow much more flexibility for the client and it will be easier for such cases.
Dynamic client is also the way all the other clients use (python, java and go).

**Comment by an-tex** [comment]
>  it will require you to create those yamls definitions ([example](https://gist.githubusercontent.com/yue9944882/266fee8e95c2f15a93778263633e72ed/raw/be12c13379eeed13d2532cb65da61fffb19ee3e7/crontab-crd.yaml))

Don't all third party resources have to define the CRDs themselves? Like in the [EventBus example](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/base/crds/argoproj.io_eventbus.yaml)

> Using dynamic client will allow much more flexibility for the client and it will be easier for such cases. Dynamic client is also the way all the other clients use (python, java and go).

But this will allow only for untyped usage right? Typed usage is the important point I'm after

**Comment by an-tex** [comment]
I think I didn't make myself clear about the usecase. I'd love to replace yaml based kustomize resources with typesafe scala based descriptions of the same and then exporting them as yaml to be applied on the cluster. But thinking about it this project is probably not the right scope for such a use case.

The idea came when I had a look at the documentation and this was already possible to define resources that are part the core k8s resources in a typesafe manner using skuber and then exporting them to yaml.

**Comment by hagay3** [comment]
> > it will require you to create those yamls definitions ([example](https://gist.githubusercontent.com/yue9944882/266fee8e95c2f15a93778263633e72ed/raw/be12c13379eeed13d2532cb65da61fffb19ee3e7/crontab-crd.yaml))
> 
> Don't all third party resources have to define the CRDs themselves? Like in the [EventBus example](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/base/crds/argoproj.io_eventbus.yaml)
> 
Yes right. not sure if the crd yaml is easy to find for every case.

> > Using dynamic client will allow much more flexibility for the client and it will be easier for such cases. Dynamic client is also the way all the other clients use (python, java and go).
> 
> But this will allow only for untyped usage right? Typed usage is the important point I'm after
Yes right, for the type case you need to provide a case class.

I will link this issue to OpenAPI integration #180, once skuber will have the integration we can use "class generator" for this use case as well.

**Comment by an-tex** [comment]
> > > it will require you to create those yamls definitions ([example](https://gist.githubusercontent.com/yue9944882/266fee8e95c2f15a93778263633e72ed/raw/be12c13379eeed13d2532cb65da61fffb19ee3e7/crontab-crd.yaml))
> > 
> > 
> > Don't all third party resources have to define the CRDs themselves? Like in the [EventBus example](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/base/crds/argoproj.io_eventbus.yaml)
> 
> Yes right. not sure if the crd yaml is easy to find for every case.

The user would have to find it manually in the corresponding github repository, but that's usually fairly easy to find. It's mostly placed in a manifests directory like with [EventBus](https://github.com/argoproj-labs/argo-eventbus/blob/main/manifests/base/crds/argoproj.io_eventbus.yaml). I guess it could be handled like a "dependency" so the user defines a custom sbt setting like libraryDependencies where the user places all the CRD urls.

> I will link this issue to OpenAPI integration #180, once skuber will have the integration we can use "class generator" for this use case as well.
 
Great thanks! :)
